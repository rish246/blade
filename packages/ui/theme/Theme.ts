/**
 * ðŸŽ¨ Theme Type â€” Manual Definition
 * ----------------------------------------------------------------------------
 * This describes the *shape* of all design tokens available in the system.
 * It is manually written to ensure type safety and flexibility, even if
 * the runtime theme objects change later.
 */

export type Theme = {
    colors: {
        bg: string; // Page background
        surface: string; // Card / elevated surfaces
        text: string; // Primary text
        muted: string; // Secondary text
        accent: string; // Brand or primary CTA
        success: string; // Success state
        warning: string; // Warning state
        error: string; // Error state
        none: string; // Empty color (no fill)
        transparent: string; // Transparent background
    };

    spacing: {
        none: string;
        xs: string;
        sm: string;
        md: string;
        lg: string;
        xl: string;
    };

    radii: {
        sm: string;
        md: string;
        lg: string;
        full: string;
    };

    typography: {
        fontFamily: string;

        // Font sizes
        fontSizeSm: string;
        fontSizeMd: string;
        fontSizeLg: string;
        fontSizeXl: string;
        fontSize2xl: string;
        fontSize3xl: string;

        // Line heights
        lineHeightSm: string;
        lineHeightMd: string;
        lineHeightLg: string;
        lineHeightXl: string;
        lineHeight2xl: string;
        lineHeight3xl: string;

        // Font weights
        fontWeightNormal: string;
        fontWeightMedium: string;
        fontWeightBold: string;
    };

    shadows: {
        sm: string;
        md: string;
        lg: string;
    };
};

// -----------------------------------------------------------------------------
// ðŸ§© Token category shorthand types
// -----------------------------------------------------------------------------

export type ColorToken =
    | "bg"
    | "surface"
    | "text"
    | "muted"
    | "accent"
    | "success"
    | "warning"
    | "error"
    | "none"
    | "transparent";

export type SpacingToken = "none" | "xs" | "sm" | "md" | "lg" | "xl";

export type RadiusToken = "sm" | "md" | "lg" | "full";

export type TypographyToken =
    | "fontFamily"
    | "fontSizeSm"
    | "fontSizeMd"
    | "fontSizeLg"
    | "fontSizeXl"
    | "fontSize2xl"
    | "fontSize3xl"
    | "lineHeightSm"
    | "lineHeightMd"
    | "lineHeightLg"
    | "lineHeightXl"
    | "lineHeight2xl"
    | "lineHeight3xl"
    | "fontWeightNormal"
    | "fontWeightMedium"
    | "fontWeightBold";

export type ShadowToken = "sm" | "md" | "lg";

// -----------------------------------------------------------------------------
// ðŸŒ— Theme modes
// -----------------------------------------------------------------------------

export type ThemeMode = "light" | "dark";
