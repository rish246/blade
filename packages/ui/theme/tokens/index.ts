export { darkTheme } from "./dark";
export { lightTheme } from "./light";
export { baseTheme } from "./base";
