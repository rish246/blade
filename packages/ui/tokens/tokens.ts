export const tokens = {
    colors: {
        bg: "#ffffff", // page background
        surface: "#f9fafb", // card / elevated surfaces
        text: "#111827", // primary text
        muted: "#6b7280", // secondary text
        accent: "#2563eb", // brand / CTA buttons
        success: "#10b981", // success state
        warning: "#f59e0b", // warning state
        error: "#ef4444", // error state
        none: "",
        transparent: "transparent",
    },

    spacing: {
        xs: "4px",
        sm: "8px",
        md: "16px",
        lg: "24px",
        xl: "32px",
    },

    radii: {
        sm: "4px",
        md: "8px",
        lg: "16px",
        full: "9999px",
    },

    typography: {
        fontFamily: "'Inter', sans-serif",

        // Font sizes - now includes heading sizes
        fontSizeSm: "14px",
        fontSizeMd: "16px",
        fontSizeLg: "20px",
        fontSizeXl: "24px", // ✅ Added for h3
        fontSize2xl: "32px", // ✅ Added for h2
        fontSize3xl: "40px", // ✅ Added for h1

        // Line heights - now includes heading line heights
        lineHeightSm: "20px",
        lineHeightMd: "24px",
        lineHeightLg: "28px",
        lineHeightXl: "32px", // ✅ Added for h3
        lineHeight2xl: "40px", // ✅ Added for h2
        lineHeight3xl: "48px", // ✅ Added for h1

        // Font weights
        fontWeightNormal: "400",
        fontWeightMedium: "500",
        fontWeightBold: "700",
    },

    shadows: {
        sm: "0 1px 2px rgba(0, 0, 0, 0.05)",
        md: "0 4px 6px rgba(0, 0, 0, 0.1)",
        lg: "0 10px 15px rgba(0, 0, 0, 0.15)",
    },
} as const;

// Types for autocompletion
export type ColorToken = keyof typeof tokens.colors;
export type SpacingToken = keyof typeof tokens.spacing;
export type RadiusToken = keyof typeof tokens.radii;
export type TypographyToken = keyof typeof tokens.typography;
export type ShadowToken = keyof typeof tokens.shadows;
export type TokenCategory = keyof typeof tokens;
