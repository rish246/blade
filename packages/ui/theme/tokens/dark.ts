import { Theme } from "../Theme";
import { baseTheme } from "./base";

/**
 * ðŸŒš Dark Theme
 * --------------------------------------------------------------------------
 * A deep, high-contrast variant optimized for low-light environments.
 * Extends the base theme with dark-mode specific color values.
 */
export const darkTheme: Theme = {
    ...baseTheme,

    colors: {
        ...baseTheme.colors,

        // Core backgrounds
        bg: "#0f172a", // Slate-900 â€” main background
        surface: "#1e293b", // Slate-800 â€” elevated surfaces

        // Text
        text: "#f1f5f9", // Light text
        muted: "#94a3b8", // Muted / secondary text

        // Brand / accent
        accent: "#3b82f6", // Blue-500

        // Semantic states
        success: "#22c55e", // Green-500
        warning: "#facc15", // Yellow-400
        error: "#f87171", // Red-400

        none: "",
        transparent: "transparent",
    },
} as const;

export type DarkTheme = typeof darkTheme;
