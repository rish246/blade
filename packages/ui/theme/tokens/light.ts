import { Theme } from "../Theme";
import { baseTheme } from "./base";

/**
 * ðŸŒž Light Theme
 * --------------------------------------------------------------------------
 * The default color scheme â€” bright surfaces, dark text, soft shadows.
 * Extends the base theme with light-mode specific color values.
 */
export const lightTheme: Theme = {
    ...baseTheme,

    colors: {
        ...baseTheme.colors,

        // Core backgrounds
        bg: "#ffffff", // Page background
        surface: "#f9fafb", // Card / surfaces

        // Text
        text: "#111827", // Primary text (near-black)
        muted: "#6b7280", // Secondary text

        // Brand / accent
        accent: "#2563eb", // Blue-600 (vivid brand color)
        // Semantic states
        success: "#10b981", // Green-500
        warning: "#f59e0b", // Amber-500
        error: "#ef4444", // Red-500

        // Utility

        none: "",
        transparent: "transparent",
    },
} as const;

export type LightTheme = typeof lightTheme;
