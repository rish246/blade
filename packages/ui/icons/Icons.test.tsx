import { render, screen } from "@testing-library/react";
import { vi } from "vitest";
import Icon from ".";
import { baseTheme } from "../theme/tokens";

describe("Icon Component", () => {
    it("renders the correct icon based on name", () => {
        render(<Icon name="Search" data-testid="icon" />);
        const wrapper = screen.getByTestId("icon");
        expect(wrapper).toBeInTheDocument();

        // Check for SVG or Lucide symbol (SVGR output may vary)
        const svg = wrapper.firstChild;
        expect(svg).not.toBeNull();
    });

    it("renders with default size and color", () => {
        render(<Icon name="User" />);
        const wrapper = screen.getByRole("img", { hidden: true });
        const svg = wrapper.querySelector("svg");
        expect(svg).toHaveAttribute("width", "24");
        expect(svg).toHaveAttribute("height", "24");
        expect(svg).toHaveAttribute("stroke", baseTheme.colors.accent);
    });

    it("applies custom size and color props", () => {
        render(<Icon name="Check" size={32} color="red" />);
        const wrapper = screen.getByRole("img", { hidden: true });
        const svg = wrapper.querySelector("svg");
        expect(svg).toHaveAttribute("width", "32");
        expect(svg).toHaveAttribute("height", "32");
        expect(svg).toHaveAttribute("stroke", "red");
    });

    it("renders with custom className", () => {
        const { container } = render(<Icon name="Menu" className="my-icon" />);
        expect(container.firstChild).toHaveClass("my-icon");
    });

    it("merges custom styles with icon color", () => {
        const { container } = render(
            <Icon name="Info" style={{ marginTop: "10px", color: "blue" }} />,
        );
        const wrapper = container.firstChild;
        expect(wrapper).toHaveStyle({ marginTop: "10px" });

        // Ensure SVG stroke color reflects style
        const svg = wrapper.firstChild;
        expect(svg).toHaveAttribute("stroke", "blue");
    });

    // ==================== Accessibility ====================
    describe("Accessibility", () => {
        it("is aria-hidden by default (decorative icon)", () => {
            render(<Icon name="Heart" />);
            const wrapper = screen.getByRole("img", { hidden: true });
            expect(wrapper).toHaveAttribute("aria-hidden", "true");
        });

        it("renders with aria-label when provided", () => {
            render(<Icon name="TriangleAlert" aria-label="Warning" />);
            const wrapper = screen.getByLabelText("Warning");
            expect(wrapper).toBeInTheDocument();
            expect(wrapper).toHaveAttribute("role", "img");
            expect(wrapper).toHaveAttribute("aria-hidden", "false");
        });

        it("overrides aria-hidden when explicitly set", () => {
            render(
                <Icon name="User" aria-hidden={false} aria-label="User Icon" />,
            );
            const wrapper = screen.getByLabelText("User Icon");
            expect(wrapper).toHaveAttribute("aria-hidden", "false");
        });

        it("sets role='img' for screen readers", () => {
            render(<Icon name="Info" />);
            const wrapper = screen.getByRole("img", { hidden: true });
            expect(wrapper).toBeInTheDocument();
        });

        it("applies focusable='false' to the inner svg", () => {
            render(<Icon name="Trash" />);
            const wrapper = screen.getByRole("img", { hidden: true });
            const svg = wrapper.querySelector("svg");
            expect(svg).toHaveAttribute("focusable", "false");
        });
    });

    // ==================== Invalid Icons ====================
    describe("Invalid Icons", () => {
        it("returns null and warns when invalid icon name is passed", () => {
            const spy = vi.spyOn(console, "warn").mockImplementation(() => {});
            const { container } = render(<Icon name={"Nonexistent" as any} />);
            expect(container.firstChild).toBeNull();
            expect(spy).toHaveBeenCalledWith(
                expect.stringContaining("⚠️ Icon not found: Nonexistent"),
            );
            spy.mockRestore();
        });
    });
});
