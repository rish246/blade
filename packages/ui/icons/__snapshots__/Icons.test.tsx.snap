// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Icon Component > Snapshots > matches snapshot for default icon 1`] = `<DocumentFragment />`;

exports[`Icon Component > Snapshots > matches snapshot with aria-label and custom styles 1`] = `<DocumentFragment />`;
